# The U+FFFD copy-corruption bug in `TokTrie`

**Status: found, root-caused, and fixed on 2026-08-18.**

Symptom that started this: on UniversalNER, `gpt-oss-20b` with constrained decoding
produced a non-zero `wrong_text_rate` on the Chinese treebanks. One example
(`zh_pud:377`) came out as

```
... 但 <SPAN><LABEL>PER</LABEL>蓋耶</SPAN><SPAN><LABEL>LOC</LABEL>�</SPAN> 已經 到 了 ...
```

with a literal replacement character (U+FFFD, `�`) inside a span that the model's own
reasoning trace never asked for.

**The bug is NOT in the FSM / logits processor.** It is in how `TokTrie` records each
token's bytes. The processor faithfully enforced "copy the input" against a *copy source
that was already corrupted before generation started*.

Sections below:
1. The one-line cause
2. How much of each vocabulary is affected
3. The failing sentence, end to end
4. Why `gpt-oss` broke and `gemma-4-E2B` did not (and why gemma is **not** immune)
5. The fix, verified


## 1. The one-line cause

`build_toktrie_from_tokenizer` recovered each token's bytes like this:

```python
surface = tokenizer.decode([token_id], clean_up_tokenization_spaces=False)
token_bytes = surface.encode("utf-8")
```

`decode()` of a **single** token that is only *part* of a multi-byte character cannot
return a valid string, so it returns `�`. The following `.encode("utf-8")` then
faithfully encodes that replacement character as `b'\xef\xbf\xbd'`.

The token's real bytes are destroyed and replaced by U+FFFD's bytes. The round-trip is
lossy exactly where it matters.


In [1]:
import sys
sys.path.insert(0, "../../src")
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained("openai/gpt-oss-20b")

# '蓋' is 3 bytes in UTF-8 and gpt-oss splits it across TWO tokens,
# neither of which is a valid character on its own.
ch = "蓋"
ids = tok.encode(ch, add_special_tokens=False)
print(f"{ch!r} = {ch.encode('utf-8')!r}  ->  {len(ids)} tokens: {ids}")

for tid in ids:
    surface = tok.decode([tid], clean_up_tokenization_spaces=False)
    print(f"  id={tid:<6} token_str={tok.convert_ids_to_tokens(tid)!r:<8} "
          f"decode()={surface!r}  decode().encode('utf-8')={surface.encode('utf-8')!r}")

print()
print("The two tokens SHOULD carry b'\\xe8\\x93' and b'\\x8b'.")
print("Both instead get recorded as b'\\xef\\xbf\\xbd' -- the bytes of U+FFFD.")

'蓋' = b'\xe8\x93\x8b'  ->  2 tokens: [85996, 233]
  id=85996  token_str='èĵ'     decode()='�'  decode().encode('utf-8')=b'\xef\xbf\xbd'
  id=233    token_str='ĭ'      decode()='�'  decode().encode('utf-8')=b'\xef\xbf\xbd'

The two tokens SHOULD carry b'\xe8\x93' and b'\x8b'.
Both instead get recorded as b'\xef\xbf\xbd' -- the bytes of U+FFFD.


In [2]:
from utils.TokTrie import TokTrie

def build_naive_toktrie(tokenizer):
    """Frozen copy of the ORIGINAL (buggy) build_toktrie_from_tokenizer, for demonstration."""
    naive = TokTrie()
    for token_id in tokenizer.get_vocab().values():
        surface = tokenizer.decode([token_id], clean_up_tokenization_spaces=False)
        token_bytes = surface.encode("utf-8")
        naive.insert(token_bytes, token_id)
    return naive


## 2. How much of each vocabulary is corrupted

Any token whose recorded bytes contain `b'\xef\xbf\xbd'` is suspect. (A handful of
tokens legitimately *are* the replacement character, but they are a rounding error next
to these counts.)


In [3]:
from utils.TokTrie import build_toktrie_from_tokenizer

FFFD = "\ufffd".encode("utf-8")   # b'\xef\xbf\xbd'

for model_id in ["openai/gpt-oss-20b", "google/gemma-4-E2B-it"]:
    t = AutoTokenizer.from_pretrained(model_id)
    trie = build_naive_toktrie(t)
    bad = {tid for tid, b in trie.token_id_to_bytes.items() if FFFD in b}
    print(f"{model_id:<26} corrupted: {len(bad):>5} / {len(trie.token_id_to_bytes):>6} "
          f"({100*len(bad)/len(trie.token_id_to_bytes):.2f}%)")
    for tid in sorted(bad)[:4]:
        print(f"    id={tid:<7} token_str={t.convert_ids_to_tokens(tid)!r}")
    print()

print("=" * 60)
print("For comparison, if we use the correct build_toktrie_from_tokenizer, we get no corruption only the replacement bytes:")
for model_id in ["openai/gpt-oss-20b", "google/gemma-4-E2B-it"]:
    t = AutoTokenizer.from_pretrained(model_id)
    trie = build_toktrie_from_tokenizer(t)
    bad = {tid for tid, b in trie.token_id_to_bytes.items() if FFFD in b}
    print(f"{model_id:<26} corrupted: {len(bad):>5} / {len(trie.token_id_to_bytes):>6} "
          f"({100*len(bad)/len(trie.token_id_to_bytes):.2f}%)")
    for tid in sorted(bad)[:4]:
        print(f"    id={tid:<7} token_str={t.convert_ids_to_tokens(tid)!r}")
    print()


openai/gpt-oss-20b         corrupted:  1584 / 200019 (0.79%)
    id=94      token_str='¡'
    id=95      token_str='¢'
    id=96      token_str='£'
    id=97      token_str='¤'

google/gemma-4-E2B-it      corrupted:   134 / 262144 (0.05%)
    id=366     token_str='<0x80>'
    id=367     token_str='<0x81>'
    id=368     token_str='<0x82>'
    id=369     token_str='<0x83>'

For comparison, if we use the correct build_toktrie_from_tokenizer, we get no corruption only the replacement bytes:
openai/gpt-oss-20b         corrupted:    22 / 200019 (0.01%)
    id=3251    token_str='�'
    id=10123   token_str='��'
    id=27877   token_str='����'
    id=28151   token_str='Ġ�'

google/gemma-4-E2B-it      corrupted:     6 / 262144 (0.00%)
    id=38914   token_str='��'
    id=58588   token_str='▁�'
    id=89854   token_str='����'
    id=196995  token_str='��������'



## 3. The failing sentence, end to end

`input_token_bytes` -- the copy source the FSM constrains the model against -- is built
by looking every input token up in `token_id_to_bytes`. So the corruption is present in
the processor's view of the input *before a single token is generated*.

If concatenating those recorded bytes does not reproduce `input_text.encode("utf-8")`,
verbatim copying is impossible by construction.


In [4]:
sentence = ("《 真的 真的 很愛 你 》 8 月 發布 ， 打入 了 前 50 ； 相關 單曲 《 立足 點 》 "
            "九 月 發布 ， 但 蓋耶 已經 到 了 美國 。")

trie = build_naive_toktrie(tok)
bad = {tid for tid, b in trie.token_id_to_bytes.items() if FFFD in b}

ids = tok.encode(sentence, add_special_tokens=False)
rebuilt = b"".join(trie.token_id_to_bytes[i] for i in ids)
truth = sentence.encode("utf-8")

print("copy source == real input ?", rebuilt == truth)
print(f"len(rebuilt)={len(rebuilt)}  len(truth)={len(truth)}")
print(f"tokens={len(ids)}  corrupted tokens used={sum(1 for i in ids if i in bad)}")
print()

offset = 0
for i in ids:
    b = trie.token_id_to_bytes[i]
    if i in bad:
        print(f"first culprit: id={i} token_str={tok.convert_ids_to_tokens(i)!r}")
        print(f"  recorded bytes : {b!r}")
        print(f"  TRUE text there: {truth[offset:offset+len(b)+3]!r}")
        break
    offset += len(b)


print()
print("=" * 60)
print("For comparison, the REAL (current, patched) builder:")
from utils.TokTrie import build_toktrie_from_tokenizer
fixed_trie = build_toktrie_from_tokenizer(tok)
fixed_rebuilt = b"".join(fixed_trie.token_id_to_bytes[i] for i in ids)
print("  copy source == real input ?", fixed_rebuilt == truth, "  <- what ships today")


copy source == real input ? False
len(rebuilt)=172  len(truth)=160
tokens=57  corrupted tokens used=7

first culprit: id=81542 token_str='Ġåĸ'
  recorded bytes : b' \xef\xbf\xbd'
  TRUE text there: b' \xe5\x96\xae\xe6\x9b\xb2'

For comparison, the REAL (current, patched) builder:
  copy source == real input ? True   <- what ships today


## 4. Why `gpt-oss` broke and `gemma-4-E2B` did not

The two models use different vocabulary families, and they fail under different inputs.

| | family | CJK representation | corrupted entries |
|---|---|---|---|
| `gpt-oss-20b` | byte-level BPE (o200k / tiktoken) | often **split mid-character** | 1,584 / 200,019 |
| `gemma-4-E2B-it` | SentencePiece + byte fallback | whole characters in vocab | 134 / 262,144 |

* **gpt-oss** stores byte-level BPE tokens, so a token boundary can fall *inside* a
  character. `蓋` becomes `'èĵ'` + `'ĭ'` -- neither decodes on its own.
* **gemma** has the common CJK characters as whole tokens, so each one decodes cleanly
  and the round-trip happens to survive. Its 134 corrupted entries are the `<0xNN>`
  byte-fallback tokens, which this text never uses.


In [6]:
gemma = AutoTokenizer.from_pretrained("google/gemma-4-E2B-it")

for ch in ["單", "已", "蓋"]:
    g = gemma.encode(ch, add_special_tokens=False)
    o = tok.encode(ch, add_special_tokens=False)
    print(f"{ch!r} {ch.encode('utf-8')!r}")
    print(f"   gemma  : {len(g)} token(s) {[gemma.convert_ids_to_tokens(c) for c in g]}")
    print(f"   gpt-oss: {len(o)} token(s) {[tok.convert_ids_to_tokens(c) for c in o]}")


'單' b'\xe5\x96\xae'
   gemma  : 1 token(s) ['單']
   gpt-oss: 1 token(s) ['åĸ®']
'已' b'\xe5\xb7\xb2'
   gemma  : 1 token(s) ['已']
   gpt-oss: 1 token(s) ['å·²']
'蓋' b'\xe8\x93\x8b'
   gemma  : 1 token(s) ['蓋']
   gpt-oss: 2 token(s) ['èĵ', 'ĭ']


### gemma is **latently** affected, not immune

The distinction matters for the write-up: gemma survived *this data*, not the bug. As
soon as a character falls outside its vocabulary it drops to `<0xNN>` byte fallback --
and every one of those tokens is recorded as U+FFFD, so the copy invariant breaks in
exactly the same way.


In [7]:
gtrie = build_naive_toktrie(gemma)
gbad = {tid for tid, b in gtrie.token_id_to_bytes.items() if FFFD in b}

samples = [
    ("common CJK", "已經 到 了 美國"),
    ("emoji",      "I love 🍕 pizza"),
    ("rare CJK",   "𠮷野家 restaurant"),
    ("math",       "x ⨌ y ⫘ z"),
    ("Deseret",    "𐐀𐐁𐐂 script"),
]

for name, text in samples:
    tids = gemma.encode(text, add_special_tokens=False)
    rebuilt = b"".join(gtrie.token_id_to_bytes[i] for i in tids)
    n_fallback = sum(1 for i in tids if i in gbad)
    ok = rebuilt == text.encode("utf-8")
    print(f"[{'OK    ' if ok else 'BROKEN'}] {name:<12} {text!r}")
    print(f"           byte-fallback tokens used: {n_fallback}")


[OK    ] common CJK   '已經 到 了 美國'
           byte-fallback tokens used: 0
[OK    ] emoji        'I love 🍕 pizza'
           byte-fallback tokens used: 0
[BROKEN] rare CJK     '𠮷野家 restaurant'
           byte-fallback tokens used: 4
[BROKEN] math         'x ⨌ y ⫘ z'
           byte-fallback tokens used: 6
[BROKEN] Deseret      '𐐀𐐁𐐂 script'
           byte-fallback tokens used: 12


## 5. The fix

The instinct: **never round-trip through `str`.** But there are two wrong ways
to act on that instinct, and both fail silently. Section 5 walks through both before
giving the version that is actually correct.

### Wrong turn 1 -- `convert_ids_to_tokens(...).encode("utf-8")`

The obvious replacement for `decode()` is `convert_ids_to_tokens()`, then encode the
result. This is *also* lossy, for a different reason.

In a byte-level BPE vocabulary the token *string* is already a re-encoding of raw bytes
into printable Unicode: the character `'e`-with-grave` **stands for** the single byte
`0xE8`. It is a stand-in symbol, not text to be encoded. Encoding it produces the UTF-8
bytes of the *symbol*, which is a completely different byte string.


In [8]:
tok = AutoTokenizer.from_pretrained("openai/gpt-oss-20b")

ids = tok.encode("蓋", add_special_tokens=False)
truth = "蓋".encode("utf-8")
print(f"'蓋' true bytes: {truth!r}   ({len(truth)} bytes)")

for t in ids:
    s = tok.convert_ids_to_tokens(t)
    print(f"  id={t:<7} token_str={s!r:<6} .encode('utf-8')={s.encode('utf-8')!r}")

naive = b"".join(tok.convert_ids_to_tokens(t).encode("utf-8") for t in ids)
print(f"\nconcatenated: {naive!r}")
print(f"correct?      {naive == truth}")


'蓋' true bytes: b'\xe8\x93\x8b'   (3 bytes)
  id=85996   token_str='èĵ'   .encode('utf-8')=b'\xc3\xa8\xc4\xb5'
  id=233     token_str='ĭ'    .encode('utf-8')=b'\xc4\xad'

concatenated: b'\xc3\xa8\xc4\xb5\xc4\xad'
correct?      False


### Wrong turn 2 -- guessing the vocabulary family per token

The natural next attempt is to invert the byte-level map, and fall back to plain UTF-8
when a character is not in that map:

```python
try:
    return bytes(_BYTE_DECODER[c] for c in s)   # byte-level BPE
except KeyError:
    return s.encode("utf-8")                    # assume SentencePiece
```

This is the dangerous one, because **the `try` branch succeeds on the wrong tokenizer**.
`bytes_to_unicode()` maps bytes `0xA1-0xFF` to the Latin-1 characters that look
identical to them, so a SentencePiece token like `'é'` *is* in the map -- and silently
resolves to one byte instead of two.

The result: it fixes Chinese on `gpt-oss` while quietly corrupting **German,
Portuguese, Nordic and Croatian** text on `gemma`. A per-token guess is not safe.


In [1]:
from transformers.convert_slow_tokenizer import bytes_to_unicode

_BYTE_DECODER = {v: k for k, v in bytes_to_unicode().items()}
gemma = AutoTokenizer.from_pretrained("google/gemma-4-E2B-it")

print("SentencePiece tokens that the byte-level map WRONGLY accepts:\n")
for probe in ["é", "ü", "š", "ć"]:
    for g in gemma.encode(probe, add_special_tokens=False):
        s = gemma.convert_ids_to_tokens(g)
        applies = all(c in _BYTE_DECODER for c in s)
        if applies:
            wrong = bytes(_BYTE_DECODER[c] for c in s)
            print(f"  token {s!r}: map applies -> {wrong!r}   but TRUTH is {s.encode('utf-8')!r}  <-- WRONG")
        else:
            print(f"  token {s!r}: map does not apply -> correctly falls through")


SentencePiece tokens that the byte-level map WRONGLY accepts:

  token 'é': map applies -> b'\xe9'   but TRUTH is b'\xc3\xa9'  <-- WRONG
  token 'ü': map applies -> b'\xfc'   but TRUTH is b'\xc3\xbc'  <-- WRONG
  token 'š': map does not apply -> correctly falls through
  token 'ć': map applies -> b'\x07'   but TRUTH is b'\xc4\x87'  <-- WRONG


In [2]:
print(_BYTE_DECODER)

{'!': 33, '"': 34, '#': 35, '$': 36, '%': 37, '&': 38, "'": 39, '(': 40, ')': 41, '*': 42, '+': 43, ',': 44, '-': 45, '.': 46, '/': 47, '0': 48, '1': 49, '2': 50, '3': 51, '4': 52, '5': 53, '6': 54, '7': 55, '8': 56, '9': 57, ':': 58, ';': 59, '<': 60, '=': 61, '>': 62, '?': 63, '@': 64, 'A': 65, 'B': 66, 'C': 67, 'D': 68, 'E': 69, 'F': 70, 'G': 71, 'H': 72, 'I': 73, 'J': 74, 'K': 75, 'L': 76, 'M': 77, 'N': 78, 'O': 79, 'P': 80, 'Q': 81, 'R': 82, 'S': 83, 'T': 84, 'U': 85, 'V': 86, 'W': 87, 'X': 88, 'Y': 89, 'Z': 90, '[': 91, '\\': 92, ']': 93, '^': 94, '_': 95, '`': 96, 'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103, 'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110, 'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117, 'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122, '{': 123, '|': 124, '}': 125, '~': 126, '¡': 161, '¢': 162, '£': 163, '¤': 164, '¥': 165, '¦': 166, '§': 167, '¨': 168, '©': 169, 'ª': 170, '«': 171, '¬': 172, '®': 

### The correct version -- detect the family once, from the decoder

The family is a property of the *tokenizer*, not of an individual token, so it must be
decided once and applied unconditionally:

* **byte-level BPE** (GPT-2 / tiktoken / o200k, i.e. `gpt-oss`) -- invert
  `bytes_to_unicode()` for every token.
* **SentencePiece** (`gemma`) -- a literal `<0xNN>` token carries exactly one raw byte;
  otherwise the token is ordinary text with `U+2581` marking a leading space.

`tokenizer.backend_tokenizer.decoder` tells us which, with no guessing.


In [15]:
from tokenizers import decoders


def _is_byte_level(tokenizer) -> bool:
    """Vocabulary family, decided ONCE from the decoder -- never guessed per token."""
    dec = getattr(getattr(tokenizer, "backend_tokenizer", None), "decoder", None)
    if dec is None:
        return False
    if isinstance(dec, decoders.ByteLevel):
        return True
    if isinstance(dec, decoders.Sequence):
        return "ByteLevel" in str(dec)
    return False


def true_token_bytes(tokenizer, token_id, byte_level):
    """A token's real bytes, with no bytes -> str -> bytes round-trip."""
    s = tokenizer.convert_ids_to_tokens(token_id)
    if s is None:
        return None
    if byte_level:
        # The token string is raw bytes re-encoded into printable unicode; invert it.
        try:
            return bytes(_BYTE_DECODER[c] for c in s)
        except KeyError:
            return s.encode("utf-8")
    # SentencePiece: '<0xNN>' byte fallback carries exactly one raw byte.
    if len(s) == 6 and s.startswith("<0x") and s.endswith(">"):
        try:
            return bytes([int(s[3:5], 16)])
        except ValueError:
            pass
    return s.replace("▁", " ").encode("utf-8")


for m in ["openai/gpt-oss-20b", "google/gemma-4-E2B-it"]:
    t = AutoTokenizer.from_pretrained(m)
    d = t.backend_tokenizer.decoder
    print(f"{m:<26} decoder={type(d).__name__:<12} byte_level={_is_byte_level(t)}")

tok = AutoTokenizer.from_pretrained("openai/gpt-oss-20b")
# tok = AutoTokenizer.from_pretrained("google/gemma-4-E2B-it")

# One special symbol demonstration:
symbol = "𠮷"  # U+20BB7, a rare CJK character
print(f"\nSpecial symbol {symbol!r} (U+20BB7) demonstration:")
print(f"  {symbol!r} UTF-8 bytes: {symbol.encode('utf-8')!r}  ({len(symbol.encode('utf-8'))} bytes)")
tokens = tok.encode(symbol, add_special_tokens=False)
print(f"  {symbol!r} -> {len(tokens)} token(s): {tokens}")
for tid in tokens:
    surface = tok.decode([tid], clean_up_tokenization_spaces=False)
    print(f"  id={tid:<6} token_str={tok.convert_ids_to_tokens(tid)!r:<8} "
          f"decode()={surface!r}  decode().encode()={surface.encode('utf-8')!r}")

# Use the true_token_bytes function to get the actual bytes for each token
print("\nTrue token bytes for each token:")
for tid in tokens:
    true_bytes = true_token_bytes(tok, tid, _is_byte_level(tok))
    print(f"  id={tid:<6} token_str={tok.convert_ids_to_tokens(tid)!r:<8} "
          f"true bytes={true_bytes!r}")

openai/gpt-oss-20b         decoder=ByteLevel    byte_level=True
google/gemma-4-E2B-it      decoder=Sequence     byte_level=False

Special symbol '𠮷' (U+20BB7) demonstration:
  '𠮷' UTF-8 bytes: b'\xf0\xa0\xae\xb7'  (4 bytes)
  '𠮷' -> 4 token(s): [478, 398, 412, 421]
  id=478    token_str='<0xF0>' decode()='�'  decode().encode()=b'\xef\xbf\xbd'
  id=398    token_str='<0xA0>' decode()='�'  decode().encode()=b'\xef\xbf\xbd'
  id=412    token_str='<0xAE>' decode()='�'  decode().encode()=b'\xef\xbf\xbd'
  id=421    token_str='<0xB7>' decode()='�'  decode().encode()=b'\xef\xbf\xbd'

True token bytes for each token:
  id=478    token_str='<0xF0>' true bytes=b'\xf0'
  id=398    token_str='<0xA0>' true bytes=b'\xa0'
  id=412    token_str='<0xAE>' true bytes=b'\xae'
  id=421    token_str='<0xB7>' true bytes=b'\xb7'


> ⚠️ **Limitation of `_is_byte_level` — a `Sequence` containing both.**
>
> The family test inspects `backend_tokenizer.decoder`. Two cases are unambiguous:
> a bare `ByteLevel` (GPT-OSS, Qwen, Llama), and Gemma's SentencePiece-shaped
> `Sequence(decoders=[Replace(\u2581 -> ' '), ByteFallback(), Fuse()])`.
>
> For a `Sequence` the check falls back to the substring test `"ByteLevel" in str(dec)`.
> That is correct for all 13 models measured here — Gemma's `Sequence` contains
> `ByteFallback`, a *different* class, so it correctly returns `False`. But it is a
> string match, not a type check, and it would answer badly for a tokenizer whose
> `Sequence` genuinely **mixes** `ByteLevel` with SentencePiece-style stages: it would
> report `True` and apply the byte-level inverse map to every token, silently
> mis-decoding the SentencePiece ones (the same failure as *wrong turn 2* above).
>
> **Before adding a model, print its decoder and check.** If a future tokenizer nests
> `ByteLevel` inside a `Sequence` alongside `Metaspace`/`ByteFallback`, this helper
> needs to walk the sequence and decide per stage rather than per tokenizer — or the
> byte recovery has to be driven by the individual stages in order.
>
> ```python
> print(type(tok.backend_tokenizer.decoder).__name__, "|", tok.backend_tokenizer.decoder)
> ```
>
> A cheap guard that catches this class of error regardless of decoder shape: assert the
> copy invariant on a sample of real text — `b"".join(true_token_bytes(...)) == text.encode("utf-8")`
> — which is exactly what the validation cell below does. It is the invariant the
> processor depends on, so testing it directly is stronger than trusting the family test.


### Validation on the full UniversalNER test data

Both tokenizer families, all 7,523 examples, comparing the old round-trip against the
corrected byte recovery. The probe strings at the end are the ones that broke earlier --
including the accented-Latin cases that *wrong turn 2* would have silently corrupted.


In [ ]:
import glob, json
from pathlib import Path

# Get the data directory from UNER
DATA = Path("data/semin/universal_ner")
assert DATA is not None, f"universal_ner not found walking up from {Path.cwd()}"

files = sorted(DATA.glob("uner_*.json"))
assert files, f"no uner_*.json under {DATA}"
corpus = [r["text"] for p in files for r in json.loads(p.read_text(encoding="utf-8"))]
print(f"data: {DATA}")
print(f"{len(files)} subsets, {len(corpus)} examples\n")

# Check all models for byte-level correctness, and count how many of the probes are correctly handled.
probes = ["\u5df2\u7d93 \u5230 \u4e86 \u7f8e\u570b", "\U00020bb7\u91ce\u5bb6 restaurant",
          "x \u2a0c y \u2ad8 z", "I love \U0001f355 pizza",
          "Gr\u00fc\u00dfe aus M\u00fcnchen", "\u041c\u043e\u0441\u043a\u0432\u0430",
          "\u010cesto \u0161e\u0107er, \u010fakujem"]

models = [
    "google/gemma-4-E2B-it",
    "google/gemma-4-E4B-it",
    "google/gemma-4-26B-A4B-it",
    "google/gemma-4-31B-it",
    "Qwen/Qwen3-8B",
    "Qwen/Qwen3.5-4B",
    "Qwen/Qwen3.5-9B",
    "Qwen/Qwen3.5-27B",
    "Qwen/Qwen3.6-35B-A3B",
    "Qwen/Qwen3.6-27B",
    "Qwen/Qwen3.8-27B",
    "openai/gpt-oss-20b",
    "openai/gpt-oss-120b",
]

print(f"{'model':<28}{'byte_lvl':<10}{'OLD broken':<12}{'NEW broken':<12}{'probes'}")
print("-" * 76)
for MODEL in models:
    try:
        t = AutoTokenizer.from_pretrained(MODEL)
    except Exception as e:
        print(f"{MODEL:<28}(tokenizer unavailable: {type(e).__name__})")
        continue
    bl = _is_byte_level(t) # Check if the tokenizer is byte-level.

    # Cache per-token bytes: decode() per occurrence is what made this cell slow.
    old_cache, new_cache = {}, {}
    # The old cache uses the tokenizer's decode() method to get the string representation of a token ID, then encodes it to bytes.
    def ob(i):
        b = old_cache.get(i)
        if b is None:
            b = old_cache[i] = t.decode([i], clean_up_tokenization_spaces=False).encode("utf-8")
        return b
    # The new cache uses the true_token_bytes() function to get the actual byte representation of a token ID, 
    # avoiding any potential corruption from the tokenizer's decode() method.
    def nb_(i):
        b = new_cache.get(i)
        if b is None:
            b = new_cache[i] = true_token_bytes(t, i, bl)
        return b

    bad_old = bad_new = 0
    for text in corpus:
        ids = t.encode(text, add_special_tokens=False)
        truth = text.encode("utf-8")
        if b"".join(ob(i) for i in ids) != truth:
            bad_old += 1
        if b"".join(nb_(i) for i in ids) != truth:
            bad_new += 1

    n_ok = sum(
        b"".join(nb_(i) for i in t.encode(p, add_special_tokens=False)) == p.encode("utf-8")
        for p in probes
    )
    print(f"{MODEL:<28}{str(bl):<10}{bad_old:<12}{bad_new:<12}{n_ok}/{len(probes)}")


data: data/semin/universal_ner
18 subsets, 7523 examples

model                       byte_lvl  OLD broken  NEW broken  probes
----------------------------------------------------------------------------
google/gemma-4-E2B-it       False     8           0           7/7
google/gemma-4-E4B-it       False     8           0           7/7
google/gemma-4-26B-A4B-it   False     8           0           7/7
google/gemma-4-31B-it       False     8           0           7/7
Qwen/Qwen3-8B               True      1380        0           7/7
Qwen/Qwen3.5-4B             True      179         0           7/7
Qwen/Qwen3.5-9B             True      179         0           7/7
Qwen/Qwen3.5-27B            True      179         0           7/7
Qwen/Qwen3.6-35B-A3B        True      179         0           7/7
Qwen/Qwen3.6-27B            True      179         0           7/7
Qwen/Qwen3.8-27B            True      179         0           7/7
openai/gpt-oss-20b          True      1223        0           7/7
open

### What this means for the results

* The FSM was always correct. It enforced verbatim copying against a copy source that
  had already been corrupted at trie-construction time.
* Any run whose text uses tokens that split mid-character had a corrupted copy source.
  For `gpt-oss` that is the CJK UniversalNER subsets; Latin-script data is largely
  unaffected because those tokens are single-byte.
* Affected runs must be re-run after the fix.
* `raw_token_ids` is now logged in every JSONL row, so a future decode-level artifact
  can be traced back to the exact token instead of being reverse-engineered from an
  already-lossy string.


# Bug 2 — tag boundaries inside a character (unmasked by the Bug 1 fix)

**Status: found 2026-08-19, after the `TokTrie` fix. A missing invariant in the FSM,
not in the trie.**

Symptom: `Qwen3-8B` still produced a non-zero `wrong_text_rate` on `zh_pud` *after*
Bug 1 was fixed:

```
INPUT : ... 但 蓋耶 已經 到 了 美國 。
OUTPUT: ... 但 蓋耶 已經 到 了 <SPAN><LABEL>LOC</LABEL>�</SPAN><SPAN><LABEL>LOC</LABEL>�</SPAN>國 。
```

**This is a second, independent bug, and the Bug 1 fix is what exposed it.**

The two are causally linked:

* Before the fix, byte-fallback tokens were filed in the trie under U+FFFD's bytes, so
  `prefix_search(b'\xe7')` returned **nothing** — partial-byte copies were unreachable.
* After the fix they are filed under their true bytes, so `prefix_search` correctly
  offers them — and the FSM had no rule preventing a tag from opening or closing
  *inside* a character.

The fix was right. It simply made a pre-existing gap reachable for the first time.


## 1. What the model actually emitted

`raw_token_ids` (added to every JSONL row while chasing Bug 1) makes this readable
directly, instead of being reverse-engineered from a lossy decoded string.

`美` is `b'\xe7\xbe\x8e'` — three bytes. The model copied it as three **separate
single-byte tokens**, and closed the span after only two of them.


In [3]:
import sys, json
from pathlib import Path

ROOT = next(p for p in [Path.cwd(), *Path.cwd().parents] if (p / "src" / "utils").is_dir())
if str(ROOT / "src") not in sys.path:
    sys.path.insert(0, str(ROOT / "src"))

from transformers import AutoTokenizer
from utils.TokTrie import build_toktrie_from_tokenizer

P = ROOT / ("Experiment_results_publication/Archived_results/UNER/"
            "Predictions/uner_all_Qwen3-8B_bug_found.jsonl")
assert P.is_file(), f"prediction file not found: {P}"

rows = [json.loads(l) for l in P.read_text(encoding="utf-8").splitlines() if l.strip()]
bad = [r for r in rows if r["wrong_text"]][0]

qwen = AutoTokenizer.from_pretrained(bad["model"])
qtrie = build_toktrie_from_tokenizer(qwen)

print("INPUT :", bad["input_text"][-14:])
print("OUTPUT:", bad["raw_output"][-70:])
print()
print("tokens 84-99 of the generation:")
for k, t in enumerate(bad["raw_token_ids"], start=1):
    if 84 <= k <= 99:
        print(f"  {k:>3} id={t:<8} {qwen.convert_ids_to_tokens(t)!r:<10} "
              f"bytes={qtrie.token_id_to_bytes.get(t)!r}")


INPUT : 蓋耶 已經 到 了 美國 。
OUTPUT:  了 <SPAN><LABEL>LOC</LABEL>�</SPAN><SPAN><LABEL>LOC</LABEL>�</SPAN>國 。

tokens 84-99 of the generation:
   84 id=29       '>'        bytes=b'>'
   85 id=163      'ç'        bytes=b'\xe7'
   86 id=122      '¾'        bytes=b'\xbe'
   87 id=522      '</'       bytes=b'</'
   88 id=67023    'SPAN'     bytes=b'SPAN'
   89 id=29       '>'        bytes=b'>'
   90 id=27       '<'        bytes=b'<'
   91 id=67023    'SPAN'     bytes=b'SPAN'
   92 id=1784     '><'       bytes=b'><'
   93 id=63290    'LABEL'    bytes=b'LABEL'
   94 id=29       '>'        bytes=b'>'
   95 id=12052    'LOC'      bytes=b'LOC'
   96 id=522      '</'       bytes=b'</'
   97 id=63290    'LABEL'    bytes=b'LABEL'
   98 id=29       '>'        bytes=b'>'
   99 id=236      'İ'        bytes=b'\x8e'


Reading the dump: tokens 85 and 86 copy `b'\xe7'` and `b'\xbe'` — bytes 1 and 2 of
`美`. Token 87 is `</`, the start of `</SPAN>`. The span closed **one byte short of a
complete character**. Byte 3 (`b'\x8e'`) then lands inside a *second* span at token 99.

Two thirds of one character in one span, the last third in another. Neither fragment is
valid UTF-8 on its own, so each renders as `�` — and `validate_reconstruction` correctly
flags the output as not matching the input.


## 2. Why this only became possible after the Bug 1 fix

`prefix_search` returns every token whose bytes are a prefix of the remaining input.
With the trie repaired, that legitimately includes partial-byte tokens.


In [4]:
for mid in ["Qwen/Qwen3-8B", "openai/gpt-oss-20b", "google/gemma-4-E2B-it"]:
    t = AutoTokenizer.from_pretrained(mid)
    tr = build_toktrie_from_tokenizer(t)
    hits = tr.prefix_search("美 。".encode("utf-8"))
    print(f"{mid}   canonical encode('美') = {t.encode('美', add_special_tokens=False)}")
    for h in sorted(hits):
        b = tr.token_id_to_bytes[h]
        note = "  <- COMPLETE character" if b == "美".encode("utf-8") else (
               "  <- partial, ends mid-character" if len(b) < 3 else "")
        print(f"    id={h:<8} bytes={b!r}{note}")
    print()


Qwen/Qwen3-8B   canonical encode('美') = [57566]
    id=163      bytes=b'\xe7'  <- partial, ends mid-character
    id=46353    bytes=b'\xe7\xbe'  <- partial, ends mid-character
    id=57566    bytes=b'\xe7\xbe\x8e'  <- COMPLETE character

openai/gpt-oss-20b   canonical encode('美') = [5084]
    id=163      bytes=b'\xe7'  <- partial, ends mid-character
    id=3904     bytes=b'\xe7\xbe'  <- partial, ends mid-character
    id=5084     bytes=b'\xe7\xbe\x8e'  <- COMPLETE character

google/gemma-4-E2B-it   canonical encode('美') = [237640]
    id=469      bytes=b'\xe7'  <- partial, ends mid-character
    id=237640   bytes=b'\xe7\xbe\x8e'  <- COMPLETE character



All three families offer partial options, so **all three are exposed** — Qwen3-8B was
simply the one that drew it.

Note these runs use `do_sample=True`. The canonical whole-character token is far more
likely, but sampling occasionally takes the byte-by-byte path. That makes this bug
**intermittent**: it varies with model, seed and sampling settings, so a clean run is
not evidence of absence.


## 3. The fix — a codepoint-boundary invariant

A tag may open or close only at a byte offset that falls **between** two complete
characters. Precompute those offsets from `input_text` itself (independent of how the
tokenizer happens to chunk it), and gate both boundary sites on them.

In `__init__`:

```python
offset = 0
self._char_boundary_offsets = {0}
for ch in input_text:
    offset += len(ch.encode("utf-8"))
    self._char_boundary_offsets.add(offset)

self._token_start_offsets: List[int] = []
_acc = 0
for b in self.input_token_bytes:
    self._token_start_offsets.append(_acc)
    _acc += len(b)
self._token_start_offsets.append(_acc)
```

Two helpers:

```python
def _global_byte_offset(self) -> int:
    return self._token_start_offsets[self.input_token_ptr] + self.input_token_byte_ptr

def _at_char_boundary(self) -> bool:
    return self._global_byte_offset() in self._char_boundary_offsets
```

Then gate the two places a tag boundary is offered in `_allowed_tokens`:

```python
# SPAN_TEXT -- closing a span
if self.span_text_has_content and self._at_char_boundary():
    allowed.add(self.SPAN_CLOSE[0])

# OUTSIDE -- opening a span
if not self._prefer_literal_angle_bracket() and self._at_char_boundary():
    ...
```

**This does not restrict sub-word tagging.** `<SPAN>P</SPAN>eter` is unaffected: `P` and
`e` are complete single-byte codepoints, so that boundary lies *between* characters. The
guard forbids only stopping *inside* one character's own bytes, where no meaningful
split exists. Combining characters (e.g. `e` + U+0301) are likewise two separate
codepoints, so a boundary between them stays legal.

**No deadlock:** all 256 single-byte tokens are present in every vocabulary checked, so
whenever the guard blocks a tag the model always has a legal continuation byte.

`self._token_start_offsets` is used to make `_global_byte_offset()` O(1) instead of O(n) when checking if the current byte offset is at a character boundary. It stores the start offsets of each token in the input text for quick global byte offset calculations.


### The two offset lists, with numbers

The guard uses two lists that are easy to confuse, because **they are computed from
different things**:

| | built from | depends on the tokenizer? |
| --- | --- | --- |
| `_char_boundary_offsets` | the characters of `input_text` | **no** — pure Unicode |
| `_token_start_offsets` | `input_token_bytes` | **yes** |

That independence is the whole point: the boundary set is ground truth about the *text*,
so it can veto whatever the tokenizer happened to do.

Worked on `美國` — 2 characters, 3 bytes each, 6 bytes total — in the case that actually
failed, where the model copied it byte by byte:


In [5]:
text = "美國"
token_bytes = [b"\xe7", b"\xbe", b"\x8e", b"\xe5", b"\x9c", b"\x8b"]   # what the model chose

# _char_boundary_offsets -- from the TEXT, tokenizer-independent
offset, char_boundaries = 0, {0}
for ch in text:
    offset += len(ch.encode("utf-8"))
    char_boundaries.add(offset)

# _token_start_offsets -- from the TOKENIZATION (trailing sentinel = total length,
# needed so the lookup is valid once all input is consumed)
acc, token_starts = 0, []
for b in token_bytes:
    token_starts.append(acc)
    acc += len(b)
token_starts.append(acc)

print(f"text={text!r}  {len(text)} chars, {len(text.encode('utf-8'))} bytes")
print(f"  _char_boundary_offsets = {sorted(char_boundaries)}   <- 2 chars x 3 bytes")
print(f"  _token_start_offsets   = {token_starts}   <- 6 tokens x 1 byte")
print(f"  token starts that are NOT legal tag positions: "
      f"{[o for o in token_starts if o not in char_boundaries]}")


text='美國'  2 chars, 6 bytes
  _char_boundary_offsets = [0, 3, 6]   <- 2 chars x 3 bytes
  _token_start_offsets   = [0, 1, 2, 3, 4, 5, 6]   <- 6 tokens x 1 byte
  token starts that are NOT legal tag positions: [1, 2, 4, 5]


Now replay what Qwen3-8B did. `_global_byte_offset()` is
`_token_start_offsets[input_token_ptr] + input_token_byte_ptr`, and the guard is simply
*"is that offset in `_char_boundary_offsets`?"*


In [6]:
print(f"{'step':<22}{'offset':>7}   {'boundary?':<10}{'</SPAN> allowed?'}")
print("-" * 62)
print(f"{'open span':<23}{0:>7}   {'yes':<10}{'--'}")
for k, b in enumerate(token_bytes[:3], start=1):          # the three bytes of 美
    off = token_starts[k]
    at_boundary = off in char_boundaries
    note = "  <-- closed HERE before the fix" if off == 2 else ""
    print(f"{'copy byte %d/3 of 美' % k:<22}{off:>7}   "
          f"{'yes' if at_boundary else 'NO':<10}"
          f"{'allowed' if at_boundary else 'BLOCKED'}{note}")

print()
print("The tokenizer may stop anywhere; a TAG may only stop where a CHARACTER does.")
print("For ASCII those coincide, which is why sub-word tagging (<SPAN>P</SPAN>eter)")
print("is completely unaffected -- every Latin character is one byte, so every token")
print("start is already a character boundary.")


step                   offset   boundary? </SPAN> allowed?
--------------------------------------------------------------
open span                    0   yes       --
copy byte 1/3 of 美          1   NO        BLOCKED
copy byte 2/3 of 美          2   NO        BLOCKED  <-- closed HERE before the fix
copy byte 3/3 of 美          3   yes       allowed

The tokenizer may stop anywhere; a TAG may only stop where a CHARACTER does.
For ASCII those coincide, which is why sub-word tagging (<SPAN>P</SPAN>eter)
is completely unaffected -- every Latin character is one byte, so every token
start is already a character boundary.


In [7]:
# All 256 single-byte tokens must exist, or the guard could trap generation.
for mid in ["Qwen/Qwen3-8B", "openai/gpt-oss-20b", "google/gemma-4-E2B-it"]:
    t = AutoTokenizer.from_pretrained(mid)
    tr = build_toktrie_from_tokenizer(t)
    missing = [b for b in range(256) if not tr.prefix_search(bytes([b]))]
    print(f"{mid:<26} single-byte tokens missing: {len(missing)}")


Qwen/Qwen3-8B              single-byte tokens missing: 0
openai/gpt-oss-20b         single-byte tokens missing: 0
google/gemma-4-E2B-it      single-byte tokens missing: 0


## 4. Regression check

Drives the real processor to a mid-character position and asserts that a tag can be
opened **only** at a character boundary.


In [8]:
import torch
from utils.TrieSpanConstrainedProcessorTokenAware import TrieSpanConstrainedProcessorTokenAware

def check_boundary_guard(model_id="Qwen/Qwen3-8B", text="美國 。"):
    t = AutoTokenizer.from_pretrained(model_id)
    tr = build_toktrie_from_tokenizer(t)
    p = TrieSpanConstrainedProcessorTokenAware(
        ["PER", "ORG", "LOC"], text, t, tr, reasoning_model=False,
        model_eos_token_id=t.eos_token_id, tokenizer_eos_token_id=t.eos_token_id)
    p.prompt_len, p.prev_len = 0, 1
    seq = [0]

    open_tok = p.label_open_blocks_nospace["LOC"][0]
    at_start = open_tok in p._allowed_tokens()

    # copy byte 1 of 3 of the first character -> now mid-codepoint
    b1 = [i for i in p._allowed_copy_tokens()
          if tr.token_id_to_bytes[i] == text[0].encode("utf-8")[:1]][0]
    seq.append(b1)
    p.__call__(torch.tensor([seq]), torch.zeros(1, len(t)))
    mid = p._allowed_tokens()
    open_mid = open_tok in mid
    close_mid = p.SPAN_CLOSE[0] in mid

    print(f"{model_id}  text={text!r}")
    print(f"  at boundary  -> tag-open offered : {at_start}   (want True)")
    print(f"  mid-codepoint-> tag-open offered : {open_mid}   (want False)")
    print(f"  mid-codepoint-> span-close offered: {close_mid}   (want False)")
    print(f"  mid-codepoint-> continuations available: {len(p._allowed_copy_tokens())}  (must be > 0)")
    ok = at_start and not open_mid and not close_mid and len(p._allowed_copy_tokens()) > 0
    print("  =>", "PASS" if ok else "FAIL")
    return ok

check_boundary_guard()


Qwen/Qwen3-8B  text='美國 。'
  at boundary  -> tag-open offered : True   (want True)
  mid-codepoint-> tag-open offered : False   (want False)
  mid-codepoint-> span-close offered: False   (want False)
  mid-codepoint-> continuations available: 1  (must be > 0)
  => PASS


True

### What this means for the results

* Bug 1 (`TokTrie` byte recovery) and Bug 2 (codepoint boundary) are **independent**;
  fixing the first is what made the second reachable. Both are needed for verbatim
  copying to hold on multi-byte text.
* Bug 2 is **sampling-dependent and intermittent**, unlike Bug 1 which was
  deterministic. Runs must be re-checked per seed, not once.
* For the paper: byte-level constrained decoding needs *both* correct token bytes *and*
  an explicit codepoint-boundary constraint. A byte-level trie is not sufficient on its
  own — "copy these bytes" permits stopping inside a character, which is never valid.


# Bug 3 — the tokenizer silently rewrites the input text (Unicode NFC)

**Status: found 2026-08-23, from a non-zero `wrong_text_rate` on WMT24 ESA Hindi.
Not in the trie and not in the FSM — the *copy target itself* was already wrong.**

Bugs 1 and 2 were both about the model being allowed to emit something it shouldn't.
This one is the opposite: the model behaves perfectly, and is still marked wrong,
because the text it was constrained to copy is not the text it gets graded against.

The processor builds its copy source by **re-encoding** the input:

```python
self.input_token_ids   = tokenizer.encode(input_text)         # <-- Qwen normalises HERE
self.input_token_bytes = [token_id_to_bytes[t] for t in self.input_token_ids]
```

Qwen's tokenizer applies **NFC** normalization; Gemma's and GPT-OSS's do not. Unicode
lets one visible character be written either precomposed (one codepoint) or decomposed
(base + combining mark), and NFC is the rule for picking one spelling. It normally
*composes* — but a set of characters sits on Unicode's **composition exclusion** list,
which NFC decomposes and is then forbidden to recompose. Devanagari's *nukta* letters
(`U+0958`–`U+095F`) are on that list.

So for Qwen, `फ़` (`U+095E`) becomes `फ` + `़` (`U+092B U+093C`): one character turns
into two, three bytes into six, and every gold offset after it shifts by one.

The cell below walks the whole thing on the real offending sentence.

In [ ]:
import sys, json, unicodedata as ud
from pathlib import Path

ROOT = next(p for p in [Path.cwd(), *Path.cwd().parents] if (p / "src" / "utils").is_dir())
if str(ROOT / "src") not in sys.path:
    sys.path.insert(0, str(ROOT / "src"))

from transformers import AutoTokenizer
from utils.TokTrie import build_toktrie_from_tokenizer

# The real offending sentence, straight out of the WMT24 ESA file.
rows = json.loads((ROOT / "data/semin/wmt/wmt-en-hi-social.json").read_text(encoding="utf-8"))
TEXT = next(r["text"] for r in rows if ud.normalize("NFC", r["text"]) != r["text"])

print("1) THE INPUT, exactly as stored in the dataset")
print(f"   {TEXT[:52]} ...")
print(f"   {len(TEXT)} chars, {len(TEXT.encode())} bytes")

print("\n2) WHAT THE PROCESSOR ACTUALLY FORCES THE MODEL TO COPY")
print("   input_text -> tokenizer.encode() -> each token's bytes -> glued back together")
sources = {}
for mid in ["Qwen/Qwen3-8B", "google/gemma-4-E2B-it"]:
    tk = AutoTokenizer.from_pretrained(mid)
    trie = build_toktrie_from_tokenizer(tk)
    ids = tk.encode(TEXT, add_special_tokens=False)
    src = b"".join(trie.token_id_to_bytes[i] for i in ids)
    sources[mid] = src
    flag = "OK" if src == TEXT.encode() else "!! DIFFERENT FROM THE INPUT"
    print(f"   {mid:24} {len(src):3d} bytes   {flag}")

print("\n3) WHERE QWEN DIVERGES")
got = sources["Qwen/Qwen3-8B"].decode("utf-8")
i = next(i for i, (a, b) in enumerate(zip(TEXT, got)) if a != b)
print(f"   at character {i}:")
print(f"     dataset   : {TEXT[i]!r}  U+{ord(TEXT[i]):04X}  {ud.name(TEXT[i])}")
print(f"     tokenizer : {got[i]!r}  U+{ord(got[i]):04X} + {got[i+1]!r} U+{ord(got[i+1]):04X}")
print("   -> ONE character silently became TWO (3 bytes -> 6).")
print(f"   -> and this is exactly NFC: got == NFC(input) ? {got == ud.normalize('NFC', TEXT)}")

print("\n4) THE CONSEQUENCE -- the model is blameless")
print("   A PERFECT model copies the copy-source above flawlessly. Stripping the tags")
print("   therefore yields the copy source, which is then compared to the ORIGINAL:")
print(f"     validate_reconstruction(copy_source, input_text) -> {got == TEXT}   => wrong_text = 1")
print("   Nothing the model did was wrong. The target it was constrained to was already")
print("   a different string from the one it gets graded against.")

print("\n5) WHY CHINESE AND JAPANESE NEVER TRIGGER THIS")
for label, cp in [("Han   U+6F22 (zh/ja)", 0x6F22),
                  ("kana  U+304C (ja)", 0x304C),
                  ("nukta U+095E (hi)", 0x095E)]:
    ch = chr(cp)   # explicit codepoints: a pasted literal may already be decomposed
    dec = ud.decomposition(ch) or "(none)"
    print(f"   {label:18} decomposition={dec:14} NFC rewrites it? {ud.normalize('NFC', ch) != ch}")
print("   Han has no decomposition at all -> nothing for NFC to do.")
print("   Kana DOES decompose, but NFC composes it straight back -> round-trip safe.")
print("   The Devanagari nukta letters sit on Unicode's COMPOSITION EXCLUSION list:")
print("   NFC pulls them apart and is forbidden from putting them back together.")

print("\n6) THE FIX -- normalise once at load time, then every tokenizer agrees")
fixed = ud.normalize("NFC", TEXT)
for mid, src in sources.items():
    tk = AutoTokenizer.from_pretrained(mid)
    ok = tk.decode(tk.encode(fixed, add_special_tokens=False)) == fixed
    print(f"   {mid:24} round-trips NFC text ? {ok}")
print("   (verified for all six models in the sweep, incl. both gpt-oss)")
print("   Gold span offsets must be remapped by the same +1-per-nukta shift,")
print("   which is why this belongs in the dataset loader and not in the processor.")


1) THE INPUT, exactly as stored in the dataset
   ps - मुझे उम्मीद है कि हम उसके उपन्यासों को पसंद करन ...
   139 chars, 351 bytes

2) WHAT THE PROCESSOR ACTUALLY FORCES THE MODEL TO COPY
   input_text -> tokenizer.encode() -> each token's bytes -> glued back together
   Qwen/Qwen3-8B            357 bytes   !! DIFFERENT FROM THE INPUT
   google/gemma-4-E2B-it    351 bytes   OK

3) WHERE QWEN DIVERGES
   at character 63:
     dataset   : 'फ़'  U+095E  DEVANAGARI LETTER FA
     tokenizer : 'फ'  U+092B + '़' U+093C
   -> ONE character silently became TWO (3 bytes -> 6).
   -> and this is exactly NFC: got == NFC(input) ? True

4) THE CONSEQUENCE -- the model is blameless
   A PERFECT model copies the copy-source above flawlessly. Stripping the tags
   therefore yields the copy source, which is then compared to the ORIGINAL:
     validate_reconstruction(copy_source, input_text) -> False   => wrong_text = 1
   Nothing the model did was wrong. The target it was constrained to was already
   a 

## The fix — normalize in the loader, assert in the processor

Two layers, because they do different jobs.

**1. The loader normalizes** (`span_datasets.py`). `_nfc_with_offset_map()` NFC-normalizes
the text and returns an index map, so `_normalize_text_and_spans()` can shift every gold
offset to match. It lives in `_character_example()` — the shared constructor Toxic Spans,
LegalQA and WMT all pass through — so a new dataset gets this without knowing it needs it.
`load_uner` calls it too (its spans are character offsets into `row["text"]`, so they must
be remapped *before* BIO conversion); `load_conll2003` normalizes per token, since gold is
already BIO and there are no offsets to move.

Per-character normalization only equals whole-string normalization when NFC does not
compose *across* a character boundary (`"e" + U+0301` → `é` merges two characters into one
and would invalidate the map). The helper verifies this and raises rather than assuming it.

**2. The processor asserts.** `b"".join(input_token_bytes) == input_text.encode()` at
construction. The loader is the fix; the assert is the tripwire for the dataset that has
not been added yet, or a tokenizer more aggressive than NFC — it fails at example 1
instead of silently corrupting a whole sweep.

Normalizing is **not** conceding the data to the tokenizer. The prompt is tokenized by the
same tokenizer, so the model never observes the un-normalized form either way — the
normalization has already happened by the time decoding starts. All that changes is
whether the gold agrees with the input the model provably received.

The cell below re-derives the whole thing from disk: **BEFORE** reads the raw WMT files
(what the loader used to hand over untouched), **AFTER** goes through `load_wmt()`.

In [ ]:
import sys, json, unicodedata as ud
from pathlib import Path

ROOT = next(p for p in [Path.cwd(), *Path.cwd().parents] if (p / "src" / "utils").is_dir())
if str(ROOT / "src") not in sys.path:
    sys.path.insert(0, str(ROOT / "src"))

from transformers import AutoTokenizer
from utils.TokTrie import build_toktrie_from_tokenizer
from utils.TrieSpanConstrainedProcessorTokenAware import TrieSpanConstrainedProcessorTokenAware
from utils.span_datasets import load_wmt, wmt_subsets

tk = AutoTokenizer.from_pretrained("Qwen/Qwen3-8B")
trie = build_toktrie_from_tokenizer(tk)
round_trips = lambda t: tk.decode(tk.encode(t, add_special_tokens=False)) == t

WMT = ROOT / "data/semin/wmt"
# BEFORE = the raw files on disk, i.e. what the loader used to hand over untouched.
before = {f"{n}:{i}": r
          for n in wmt_subsets()
          for i, r in enumerate(json.loads((WMT / f"wmt-{n}.json").read_text(encoding="utf-8")))}
# AFTER = the same corpus through load_wmt(), which now NFC-normalises.
after = {e["example_id"]: e for e in load_wmt("all")}

hit = [k for k, r in before.items() if ud.normalize("NFC", r["text"]) != r["text"]]

print("TABLE 1 -- whole corpus (867 examples, 24 subsets)\n")
print(f"{'':34}{'BEFORE':>10}{'AFTER':>10}")
rows = [
    ("texts not in NFC form",
     sum(ud.normalize("NFC", r['text']) != r['text'] for r in before.values()),
     sum(ud.normalize("NFC", e['text']) != e['text'] for e in after.values())),
    ("texts Qwen fails to round-trip",
     sum(not round_trips(r['text']) for r in before.values()),
     sum(not round_trips(e['text']) for e in after.values())),
]
for label, b, a in rows:
    print(f"{label:34}{b:>10}{a:>10}")

print("\n\nTABLE 2 -- the 3 affected examples, in detail\n")
hdr = f"{'example':22}{'chars':>12}{'bytes':>12}{'round-trips':>14}{'processor':>12}"
print(hdr); print("-" * len(hdr))
for k in hit:
    for tag, txt in (("before", before[k]["text"]), ("after ", after[k]["text"])):
        try:
            TrieSpanConstrainedProcessorTokenAware(
                ["MAJOR", "MINOR"], txt, tk, trie, tokenizer_eos_token_id=tk.eos_token_id)
            proc = "builds"
        except AssertionError:
            proc = "ASSERT!"
        print(f"{(k if tag=='before' else '')[:20]:22}"
              f"{tag+' '+str(len(txt)):>12}{len(txt.encode()):>12}"
              f"{str(round_trips(txt)):>14}{proc:>12}")
    print()

print("\nTABLE 3 -- gold spans still cover the same glyphs after remapping\n")
hdr = f"{'example':22}{'span':>6}  {'offsets before -> after':34}{'same text?':>12}"
print(hdr); print("-" * len(hdr))
allsame = True
for k in hit:
    braw, aex = before[k], after[k]
    old_spans = braw.get("spans") or braw.get("gold_spans") or []
    for j, (bs, as_) in enumerate(zip(old_spans, aex["gold_spans"])):
        b_sub = braw["text"][bs["start"]:bs["end"]]
        a_sub = aex["text"][as_["start"]:as_["end"]]
        same = ud.normalize("NFC", b_sub) == a_sub
        allsame &= same
        moved = "" if (bs["start"], bs["end"]) == (as_["start"], as_["end"]) else "  <- shifted"
        offsets = f'[{bs["start"]},{bs["end"]}) -> [{as_["start"]},{as_["end"]}){moved}'
        print(f'{(k if j == 0 else "")[:20]:22}{j:>6}  {offsets:34}{str(same):>12}')
print(f"\nevery gold span still covers the same text: {allsame}")


TABLE 1 -- whole corpus (867 examples, 24 subsets)

                                      BEFORE     AFTER
texts not in NFC form                      3         0
texts Qwen fails to round-trip             3         0


TABLE 2 -- the 3 affected examples, in detail

example                      chars       bytes   round-trips   processor
------------------------------------------------------------------------
en-hi-literary:14       before 381         983         False     ASSERT!
                        after  385         995          True      builds

en-hi-literary:18       before 189         481         False     ASSERT!
                        after  190         484          True      builds

en-hi-social:8          before 139         351         False     ASSERT!
                        after  141         357          True      builds


TABLE 3 -- gold spans still cover the same glyphs after remapping

example                 span  offsets before -> after             same text?
--

### What this means for the results

* **Scope is tiny but the class is not.** Only 3 of 8,894 examples across
  `wmt`/`universal_ner`/`multigec` are non-NFC (all Hindi: `U+095C`×4, `U+095E`×2,
  `U+095B`×1), and only the two Qwen tokenizers rewrite them. But the same list covers
  Bengali `U+09DC`, Gurmukhi `U+0A59`, Oriya `U+0B5C`, Hebrew presentation forms, and
  singletons like `U+212B`→`U+00C5`. It can also bite from the other direction: text
  written *decomposed* would be *composed* by NFC, changing the bytes just as much.
  The real precondition is simply **"the source text is not already in NFC form"**.
* **The fix belongs in the dataset loader, not in the processor or the trie.**
  `TokTrie` never sees `input_text` — the damage happens inside `tokenizer.encode()`,
  upstream of it. The processor sees the text but *not* `gold_spans`, and normalising
  is a coordinated transform of both: change the text without shifting the offsets and
  the metrics are computed against mismatched coordinates — silently plausible and
  wrong, which is worse than crashing. Only the loader owns both, so it goes in the
  shared `_character_example()` that Toxic Spans / LegalQA / WMT all pass through.
* **NFC is sufficient here, not in general.** It is a verified fixed point for all six
  models in the sweep, but a tokenizer applying *NFKC* (ligature `fi`→`fi`, `①`→`1`,
  full-width→half-width) would still break. The fully general invariant,
  `decode(encode(text)) == text`, is tokenizer-*dependent* — enforcing it per model
  would give each model a different gold text and destroy cross-model comparability.
  Hence two layers: the loader normalises (tokenizer-independent, once), and the
  processor asserts `b"".join(input_token_bytes) == input_text.encode()` at
  construction. The assert is the tripwire for the dataset that hasn't been added yet:
  it fires at example 1 instead of quietly corrupting a whole sweep.
* **For the paper:** verbatim-copy guarantees are stated over *bytes*, so they hold
  only if the byte sequence the constraint is built from is the same one the gold is
  measured against. A tokenizer that normalises breaks that equality before decoding
  even starts — worth one sentence, because it is invisible in any evaluation that
  does not compare the reconstruction to the original string.

In [1]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
import torch
import os

if 'model' in globals():
    del model
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

models = [
    "google/gemma-4-E2B-it", # can skip reasoning, add reasoning_start_marker
    "google/gemma-4-E4B-it", # can skip reasoning, add reasoning_start_marker
    "google/gemma-4-26B-A4B-it", # reasoning, reasoning_end_marker -> <channel|>
    "google/gemma-4-31B-it", # reasoning, reasoning_end_marker -> <channel|>
    "Qwen/Qwen3-8B", # reasoning, reasoning_end_marker -> </think>, end_of_sequence_marker -> <|endoftext|>
    "Qwen/Qwen3.5-4B", # reasoning, reasoning_end_marker -> </think>, end_of_sequence_marker -> <|endoftext|>
    "Qwen/Qwen3.5-9B", # same
    "Qwen/Qwen3.5-27B", # same
    "Qwen/Qwen3.6-35B-A3B", # same
    "Qwen/Qwen3.6-27B", # same
    'Qwen/Qwen3.8-27B',
    "openai/gpt-oss-20b", # reasoning, reasoning_end_marker -> <|channel|>final<|message|>, end_of_sequence_marker -> <|return|>
    "openai/gpt-oss-120b",
]

model_id = models[3]

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.float16,
)

tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=True, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    model_id, 
    # quantization_config=bnb_config, 
    device_map="auto",
    torch_dtype="auto",
    # device_map={"": 0}
    # torch_dtype=torch.float16
    )


Loading weights:   0%|          | 0/1188 [00:00<?, ?it/s]

In [5]:
tokenizer.encode("<|channel>")

[100]

In [15]:
for tok in ["<end_of_turn>", "<|im_end|>", "<|eot_id|>", "<|return|>", "<turn|>"]:
    tid = tokenizer.convert_tokens_to_ids(tok)
    print(tok, tid, tid == tokenizer.unk_token_id)

<end_of_turn> 3 True
<|im_end|> 3 True
<|eot_id|> 3 True
<|return|> 3 True
<turn|> 106 False


In [ ]:
from utils.model_reasoning_utils import extract_harmony_final_channel

message = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Hi, how are you doing?"},
]

prompt = tokenizer.apply_chat_template(
    message,
    tokenize=True,
    reasoning_effort="low",
    add_generation_prompt=True,
    return_tensors="pt",
    enable_thinking=True,
).to(model.device)

output = model.generate(**prompt, max_new_tokens=2000)

new_ids = output[0][prompt['input_ids'].shape[1]:]

In [2]:
print(tokenizer.decode(prompt['input_ids'][0], skip_special_tokens=False))
print("New IDs:")
print(tokenizer.decode(new_ids, skip_special_tokens=False))
# print(extract_harmony_final_channel(response))

NameError: name 'prompt' is not defined

In [14]:
# ===========================================================================
# reasoning_start_marker diagnostic -- 1/2: does the TEMPLATE open the block?
#
# Decides ModelSpec.reasoning_start_marker in src/utils/model_registry.py.
# Wrong in either direction is expensive:
#   * SET when the template pre-fills the opener -> the model never generates
#     it, so _reasoning_was_opened() is always False, EVERY run is classified
#     "skipped", and the raw reasoning trace is returned as the answer.
#   * None when the model must emit it -> a model that skips the thought block
#     (Gemma-4 E2B/E4B do) has its correct answer discarded as a truncation.
#
# Rule of thumb:
#   prompt ENDS with the opener      -> template opened it   -> registry None
#   opener absent from the prompt    -> model must emit it   -> registry = opener
#   opener present but not at the end-> ambiguous (e.g. Harmony lists its valid
#                                       channels in the preamble) -> run cell 2/2
# ===========================================================================
OPENER_CANDIDATES = ["<think>", "<|channel>", "<|channel|>analysis<|message|>", "<|think|>"]

def render(enable_thinking, effort=None):
    """Render the generation prompt, tolerating templates that reject kwargs."""
    kw = {"reasoning_effort": effort} if effort else {}
    try:
        return tokenizer.apply_chat_template(
            [{"role": "system", "content": "SYS"}, {"role": "user", "content": "USER"}],
            tokenize=False, add_generation_prompt=True,
            enable_thinking=enable_thinking, **kw,
        )
    except Exception as e:
        return f"<<template rejected these kwargs: {type(e).__name__}: {e}>>"

on, off = render(True), render(False)

print(f"model_id = {model_id}\n")
print("--- generation prompt, thinking ON (last 200 chars) ---")
print(repr(on[-200:]))
print()
print(f"template reacts to enable_thinking : {on != off}")
if on != off:
    print(f"  ON  tail: {on[-60:]!r}")
    print(f"  OFF tail: {off[-60:]!r}")

# reasoning_effort is a separate axis: check it changes the prompt at all, so a
# model registered with effort levels is not silently ignoring them.
for eff in ("low", "medium", "high", "xhigh"):
    r = render(True, eff)
    tag = "REJECTED" if r.startswith("<<") else ("changes prompt" if r != on else "accepted, no effect")
    print(f"  reasoning_effort={eff:6s} -> {tag}")

print("\n--- opener candidates ---")
ends_with, in_prompt = [], []
for c in OPENER_CANDIDATES:
    at_end = on.rstrip().endswith(c.rstrip())
    present = c in on
    if at_end:
        ends_with.append(c)
    elif present:
        in_prompt.append(c)
    print(f"  {c!r:34s} in prompt={str(present):5s}  at end={at_end}")

print("\n--- VERDICT ---")
if ends_with:
    print(f"  The TEMPLATE opens the block with {ends_with[0]!r}.")
    print("  -> reasoning_start_marker = None   (block is open by construction;")
    print("     a missing end marker can then only mean truncation)")
elif in_prompt:
    print(f"  {in_prompt} appear in the prompt but NOT at the end, so they are")
    print("  not the opener. Two things look like this:")
    print("    * a TRIGGER that only switches thinking on (Gemma-4's <|think|> in")
    print("      the system prompt) -- the model still emits its own opener later;")
    print("    * documentation (Harmony lists its valid channels in the preamble).")
    print("  -> inconclusive, run cell 2/2 to see what the model actually emits.")
else:
    print("  No opener in the prompt, so the MODEL has to emit one.")
    print("  -> reasoning_start_marker = the opener string, CONFIRMED by cell 2/2.")


model_id = google/gemma-4-E2B-it

--- generation prompt, thinking ON (last 200 chars) ---
'<bos><|turn>system\n<|think|>\nSYS<turn|>\n<|turn>user\nUSER<turn|>\n<|turn>model\n'

template reacts to enable_thinking : True
  ON  tail: 'm\n<|think|>\nSYS<turn|>\n<|turn>user\nUSER<turn|>\n<|turn>model\n'
  OFF tail: 'turn>system\nSYS<turn|>\n<|turn>user\nUSER<turn|>\n<|turn>model\n'
  reasoning_effort=low    -> accepted, no effect
  reasoning_effort=medium -> accepted, no effect
  reasoning_effort=high   -> accepted, no effect
  reasoning_effort=xhigh  -> accepted, no effect

--- opener candidates ---
  '<think>'                          in prompt=False  at end=False
  '<|channel>'                       in prompt=False  at end=False
  '<|channel|>analysis<|message|>'   in prompt=False  at end=False
  '<|think|>'                        in prompt=True   at end=False

--- VERDICT ---
  ['<|think|>'] appear in the prompt but NOT at the end, so they are
  not the opener. Two things look like this

In [16]:
# ===========================================================================
# reasoning_start_marker diagnostic -- 2/2: what does the MODEL actually emit?
#
# Cell 1/2 reads the template; this one reads reality. Run it on a SHORT, easy
# prompt -- that is exactly the condition under which Gemma-4 E2B/E4B skip the
# thought block and answer directly.
#
# Reproduces what utils_functions._reasoning_was_opened() does at eval time:
# it searches the GENERATED ids (not the prompt) for the opener token sequence.
# ===========================================================================
from utils.model_registry import get
from utils.model_reasoning_utils import find_answer_start

spec = get(model_id)
N_TRIES = 5          # short prompts, repeated -- skipping is stochastic
MAX_NEW = 300

msg = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Hi, how are you doing?"},
]
enc = tokenizer.apply_chat_template(
    msg, tokenize=True, add_generation_prompt=True,
    return_tensors="pt", return_dict=True, enable_thinking=True,
).to(model.device)

end_ids = tokenizer(spec.reasoning_end_marker, add_special_tokens=False).input_ids if spec.reasoning else None
print(f"model_id            = {model_id}")
print(f"end marker          = {spec.reasoning_end_marker!r} -> {end_ids}")
print(f"registry start mark = {spec.reasoning_start_marker!r}\n")

# Probe every candidate that tokenizes to something real for this tokenizer.
probes = {}
for c in OPENER_CANDIDATES:
    ids = tokenizer(c, add_special_tokens=False).input_ids
    if ids and tokenizer.decode(ids, clean_up_tokenization_spaces=False) == c:
        probes[c] = ids

opened_counts = {c: 0 for c in probes}
closed = 0
for i in range(N_TRIES):
    out = model.generate(**enc, max_new_tokens=MAX_NEW, do_sample=True, temperature=1.0)
    new_ids = out[0][enc["input_ids"].shape[1]:].tolist()
    for c, ids in probes.items():
        if find_answer_start(new_ids, ids) is not None:
            opened_counts[c] += 1
    if end_ids and find_answer_start(new_ids, end_ids) is not None:
        closed += 1
    if i == 0:
        print("--- first generation, special tokens visible (first 30 tokens) ---")
        print(repr(tokenizer.decode(new_ids[:30], skip_special_tokens=False)))
        print()

print(f"{'candidate opener':34s} generated in N/{N_TRIES}")
for c, n in opened_counts.items():
    print(f"  {c!r:32s} {n}/{N_TRIES}")
print(f"\nend marker emitted: {closed}/{N_TRIES}")

emitted = [c for c, n in opened_counts.items() if n > 0]
always = [c for c, n in opened_counts.items() if n == N_TRIES]

print("\n--- VERDICT ---")
if always:
    print(f"  Model emits {always[0]!r} on every try.")
    print(f"  -> reasoning_start_marker = {always[0]!r}")
    print("     (it CAN therefore be missing on other inputs, which is precisely")
    print("      the skip case the marker exists to detect)")
elif emitted:
    print(f"  Model emits {emitted} only SOMETIMES -- it skips the block on easy inputs.")
    print(f"  -> reasoning_start_marker = {emitted[0]!r}  (this is the Gemma-4 case)")
elif closed == N_TRIES:
    print("  No opener generated, but the end marker always is.")
    print("  -> the template opened the block: reasoning_start_marker = None")
else:
    print("  Neither an opener nor the end marker was generated reliably.")
    print("  -> do NOT guess. Print the raw generation above and identify the")
    print("     real delimiters before touching the registry; None is the safe")
    print("     default in the meantime.")


model_id            = google/gemma-4-E2B-it
end marker          = '<channel|>' -> [101]
registry start mark = '<|channel>'

--- first generation, special tokens visible (first 30 tokens) ---
'<|channel>thought\nThinking Process:\n\n1.  **Analyze the input:** The user said, "Hi, how are you doing?" This is a'

candidate opener                   generated in N/5
  '<think>'                        0/5
  '<|channel>'                     2/5
  '<|channel|>analysis<|message|>' 0/5
  '<|think|>'                      0/5

end marker emitted: 2/5

--- VERDICT ---
  Model emits ['<|channel>'] only SOMETIMES -- it skips the block on easy inputs.
  -> reasoning_start_marker = '<|channel>'  (this is the Gemma-4 case)


In [ ]:
from src.utils.TrieSpanConstrainedProcessorTokenAware import TrieSpanConstrainedProcessorTokenAware
from src.utils.system_prompts import SYSTEM_PROMPT_CONSTR_GEN
from src.utils.TokTrie import TokTrie, build_toktrie_from_tokenizer
from src.utils.model_reasoning_utils import reasoning_ended
from src.utils.utils_functions import generate_markup

REASONING_END_MARKER_STR = {
        'google/gemma-4-E4B-it': "<channel|>",
        'google/gemma-4-26B-A4B-it': "<channel|>",
        'google/gemma-4-31B-it': "<channel|>",
        'Qwen/Qwen3.5-4B': "</think>",
        'Qwen/Qwen3.5-9B': "</think>",
        'Qwen/Qwen3.5-27B': "</think>",
        'Qwen/Qwen3.6-35B-A3B': "</think>",
        'Qwen/Qwen3.6-27B': "</think>",
        'openai/gpt-oss-20b': "<|channel|>final<|message|>",
        'openai/gpt-oss-120b': "<|channel|>final<|message|>",
    }

reasoning_model = model_id in REASONING_END_MARKER_STR
reasoning_end_marker = (
    tokenizer(REASONING_END_MARKER_STR[model_id], add_special_tokens=False).input_ids
    if reasoning_model else None
)

print(f"model_id={model_id}  reasoning_model={reasoning_model}  reasoning_end_marker={reasoning_end_marker}")


input_text = "Radek was born in Prague"
ner_labels = ["PER", "LOC", "ORG", "MISC"]
toktrie=build_toktrie_from_tokenizer(tokenizer)

processor = TrieSpanConstrainedProcessorTokenAware(
    labels=ner_labels, input_text=input_text, tokenizer=tokenizer,
    toktrie=toktrie, reasoning_ended=reasoning_ended,
    reasoning_end_marker=reasoning_end_marker, reasoning_model=True,
    model_eos_token_id=model.generation_config.eos_token_id,
    tokenizer_eos_token_id=tokenizer.eos_token_id
)

gen_stats = {}
generated, num_output_tokens, generation_seconds = generate_markup(
                    model=model,
                    tokenizer=tokenizer,
                    processor=processor,
                    eval_model="constrained",
                    input_text=input_text,
                    system_prompt=SYSTEM_PROMPT_CONSTR_GEN,
                    max_new_tokens=5000,
                    do_sample=True if reasoning_model else False,
                    temperature=0.2,
                    reasoning_model=True,
                    reasoning_effort='low',
                    reasoning_end_marker=reasoning_end_marker,
                    stats_out=gen_stats,
                    repetition_penalty=1.2 if reasoning_model else None,
                )


In [ ]:
print([tid in tokenizer.all_special_ids for tid in [236820, 147570, 1798, 56878, 236813]])


[False, False, False, False, False]


In [ ]:
print(processor.output_start_index)

731


In [85]:
print(generated)

<SPAN><


In [84]:
print(gen_stats)

{'num_reasoning_tokens': 287, 'num_answer_tokens': 5, 'reasoning_text': 'thought\nThinking Process:\n\n1.  **Analyze the Input:** "Radek wasborn in Prague" (Wait, checking the provided example/input again - it seems there might have been a typo in my transcription if I assume standard spacing based on typical inputs, but strictly following \'Radek was borne in Prague\' structure). Assuming the actual input is "Radek wa s born in Prague". Let me use the exact string given by the user: `Radek was b orn i n P rague`. Wait, no spaces were shown between words other than those implied normally. The strict input is: `RadekwasborneinPrague`? No, usually these prompts maintain readability unless specified otherwise. Standard interpretation suggests: "Radek", "was born in", "Prague".\n\n2.  **Identify Potential Entities & Labels:**\n    *   Radek: Name of a person -> PER\n    *    Prague: City name -> LOC\n\n3.  **Apply Tagging Rules (Order: PER, LOC):**\n    * Radek -> `<SPAN><LABEL><PER></LABE